presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java

```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
+     * and before first read.
+     *
+     * All instances share visitedPositions array, but it is safe because each thread
+     * starts with visitedPositions filled with false values and marks only some positions
+     * to true. Since we don't care what will be the order of those writes to
+     * visitedPositions, writes can be without synchronization.
+     *
+     * Synchronization in commit() is required because each writing thread must have
+     * memory barrier, so that it writes its state of visitedPositions to main memory
+     * before visitedPositions is read in getVisitedPositions().
+     */
     private static class OuterPositionTracker
     {
-        private final List<Supplier<LookupSource>> lookupSourceSuppliers;
+        public static class Factory
+        {
+            private final boolean[][] visitedPositions;
+            private AtomicLong referenceCount = new AtomicLong();
+
+            public Factory(List<Integer> positionCounts)
+            {
+                visitedPositions = new boolean[positionCounts.size()][];
+                for (int partition = 0; partition < visitedPositions.length; partition++) {
+                    visitedPositions[partition] = new boolean[positionCounts.get(partition)];
+                }
+            }
 
-        @GuardedBy("this")
-        private final boolean[][] visitedPositions;
+            public synchronized OuterPositionTracker create()
+            {
+                return new OuterPositionTracker(visitedPositions, referenceCount);
+            }
+        }
 
-        @GuardedBy("this")
-        private boolean finished;
+        private final boolean[][] visitedPositions; // shared across multiple operators/drivers
+        private final AtomicLong referenceCount; // shared across multiple operators/drivers
+        private boolean written; // unique per each operator/driver
 
-        public OuterPositionTracker(List<Supplier<LookupSource>> lookupSourceSuppliers)
+        private OuterPositionTracker(boolean[][] visitedPositions, AtomicLong referenceCount)
         {
-            this.lookupSourceSuppliers = lookupSourceSuppliers;
-            visitedPositions = new boolean[lookupSourceSuppliers.size()][];
-            for (int source = 0; source < lookupSourceSuppliers.size(); source++) {
-                try (LookupSource lookupSource = lookupSourceSuppliers.get(source).get()) {
-                    visitedPositions[source] = new boolean[lookupSource.getJoinPositionCount()];
-                }
-            }
+            this.visitedPositions = visitedPositions;
+            this.referenceCount = referenceCount;
         }
 
-        public synchronized void positionVisited(int partition, int position)
+        /**
+         * No synchronization here, because it would be very expensive. Check comment above.
+         */
+        public void positionVisited(int partition, int position)
         {
-            verify(!finished);
+            if (!written) {
```

Comment:
AFAIK that was initial approach but there are cases when tracker is created but `commit()` never called, when tracker is not actually used for updating `visitedPositions` array